pkgs/build-support/rust/default.nix

```diff
@@ -117,7 +118,9 @@ stdenv.mkDerivation ((removeAttrs args ["depsExtraArgs"]) // lib.optionalAttrs u
     rustc
   ];
 
-  buildInputs = buildInputs ++ lib.optional stdenv.hostPlatform.isMinGW windows.pthreads;
+  buildInputs = buildInputs
+    ++ lib.optionals stdenv.hostPlatform.isDarwin [ libiconv ]
```

Comment:
After the update, I tried to build `fd` and `ripgrep`. They both failed on missing libiconv error. Rather than adding `libiconv` to `buildInputs` of thousands of individual rust packages, I feel this is the better way to go.

Reply:
I tend to agree if this is a dependency of the rust stdlib, otherwise I would keep it rather explicit than bloating the default closure size.. What part in `fd` and `ripgrep` depends on `libiconv`?

Reply:
But let's ask some darwin maintainer: @LnL7 @thefloweringash @abathur

Reply:
I haven't been able to pinpoint it, but I think it's clap.
I've made the fix to about 10 packages so far. I don't mind, the only problem is that for some of them, people don't spot the breakage, until somebody running on darwin comes along.

Reply:
Prior discussion: https://github.com/NixOS/nixpkgs/pull/95556

Reply:
I am trying to figure out which fd/ripgrep dependencies require libiconv and trying to fix it via addition to pkgs/build-support/rust/default-crate-overrides.nix but I fail. Any hints on what to do?

Reply:
These crate overrides are for `buildRustCrate` (e.g. `crate2nix`) and not `buildRustPackage`.

Reply:
```
$ rg -l "buildRustPackage" | grep "\.nix" |  wc -l                              
442
$ rg -l "buildRustPackage" | grep "\.nix" | xargs -n1 grep -L "libiconv" | wc -l
363
```

And among the delta are probably packages that do not compile because `libiconv` is missing. But I think it's up to @NixOS/darwin-maintainers to make this call.

Reply:
@danieldk I have fixed many Rust packages by just adding `libiconv`. I was actually planning to automatically go through Rust packages failing to build because of a lack of it.

I am in favor of making this default on darwin.

Reply:
@siraben seems reasonable then. Though I think it would be nice to find the root cause, just so that the addition of this dependency can be properly documented.

I don’t have a Mac, so I have no clue ;).

Reply:
I won't pretend to really grok the issues here, but I was looking around for context on a PR about a month ago and found these #37056 #37012

Reply:
I think darwin support can be done in another PR. AFAIK, this would have been an existing issue, and I would rather not delay this PR. Darwin maintainers have around 5 days to come up with a solution before the final staging-next cycle will begin before branch-off of 21.05

Reply:
Also, I don't think having libiconv as a buildInput will bloat too much, as the likelihood of another package requiring should be very high.

Reply:
I think I have found the culprit. The `libc` crate used to unconditionally link `libiconv` on Apple platforms:

```
    if target.contains("-apple-") {
        println!("cargo:rustc-link-lib=iconv");
    }
```

`libc` is a transitive dependency of *many* widely-used crates, such as `rand`, `memchr`,  and `chrono`, so it percolates through much of the Nix ecosystem.

*However*, this dependency is now made conditional in `libc` (only needed when `libiconv` symbols are used):

https://github.com/rust-lang/libc/pull/2150

So, I propose that we revert adding libiconv as a buildinput, since it will become unnecessary as more and more crates adopt the next version of `libc`.

Reply:
@danieldk Thanks for investigating the issue - I guess we can revert the libiconv addition together with once libc 0.2.95 update once it is out.

Reply:
Proposed PR to remove the dependency #122231.